server/src/main/java/org/elasticsearch/cluster/project/ProjectDeletedListener.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.cluster.project;
+
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.metadata.ProjectId;
+import org.elasticsearch.cluster.service.ClusterService;
+
+import java.util.function.Consumer;
+
+/**
+ * Utility class to make it easy to run a block of code whenever a project is deleted (e.g. to cleanup cache entries)
+ */
+public class ProjectDeletedListener {
```

Comment:
This class doesn't do a lot, but it makes the code that uses it a lot clearer.
The 1 line in `CompositeRolesStore` becomes really obvious:
```
  new ProjectDeletedListener(this::removeProject).attach(clusterService);
```

Reply:
It's useful if the class does not already listen for cluster state changes. I guess there are probably a number of such classes.